src/common/sys_lmdb.c

```diff
@@ -0,0 +1,511 @@
+		goto out;
+	}
+
+	rc = mdb_txn_begin(ldb->db_env, NULL, 0, &ldb->db_txn);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to begin tx for sysdb: "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+
+	rc = mdb_dbi_open(ldb->db_txn, NULL, 0, &ldb->db_dbi);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		D_CRIT("Failed to open sysdb: "DF_RC"\n", DP_RC(rc));
+		goto txn_abort;
+	}
+
+
+	d_iov_set(&key, SYS_DB_MD_VER, strlen(SYS_DB_MD_VER));
+	d_iov_set(&val, &ver, sizeof(ver));
+	if (try_create) {
+		ver = SYS_DB_VERSION;
+		rc = lmm_db_upsert(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to set version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+			goto txn_abort;
+		}
+		rc = mdb_txn_commit(ldb->db_txn);
+		if (rc)
+			D_CRIT("Failed to commit version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+		goto out;
+	} else {
+		rc = lmm_db_fetch(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to read sysdb version: "DF_RC"\n",
+			       DP_RC(rc));
+			rc = -DER_INVAL;
+		}
+
+		if (ver < SYS_DB_VERSION_1 || ver > SYS_DB_VERSION)
+			rc = -DER_DF_INCOMPT;
+	}
+txn_abort:
+	mdb_txn_abort(ldb->db_txn);
+out:
+	ldb->db_txn = NULL;
+	return rc;
+}
+
+#define MAX_SMD_TABLE_LEN	32
+static int
+lmm_db_generate_key(char *table, d_iov_t *key, MDB_val *db_key)
+{
+	char	*new_key;
+	int	 table_len;
+
+	table_len = strnlen(table, MAX_SMD_TABLE_LEN);
+	db_key->mv_size = key->iov_len + table_len;
+	D_ALLOC(new_key, db_key->mv_size);
+	if (new_key == NULL)
+		return -DER_NOMEM;
+
+	memcpy(new_key, table, table_len);
+	memcpy(new_key + table_len, (char *)key->iov_buf, key->iov_len);
+	db_key->mv_data = new_key;
+
+	return 0;
+}
+
+static int
+lmm_db_unpack_key(char *table, d_iov_t *key, MDB_val *db_key)
+{
+	int 	table_len = strnlen(table, MAX_SMD_TABLE_LEN);
+	char   *buf;
+	int	len;
+
+	if (db_key->mv_size < table_len)
+		return -DER_INVAL;
+
+	len = db_key->mv_size - table_len;
+	D_ALLOC(buf, len);
+	if (buf == NULL)
+		return -DER_NOMEM;
+
+	memcpy(buf, db_key->mv_data + table_len, len);
+	d_iov_set(key, buf, len);
+
+	return 0;
+}
+
+static int
+lmm_db_fetch(struct sys_db *db, char *table, d_iov_t *key, d_iov_t *val)
+{
+	MDB_val			 db_key, db_data;
+	struct lmm_sys_db	*ldb = db2lmm(db);
+	int			 rc;
+	bool			 end_tx = false;
+
+	if (ldb->db_txn == NULL) {
+		rc = mdb_txn_begin(ldb->db_env, NULL, MDB_RDONLY, &ldb->db_txn);
+		if (rc)
+			return mdb_error2daos_error(rc);
+		end_tx = true;
+	}
+
+	rc = lmm_db_generate_key(table, key, &db_key);
+	if (rc)
+		goto out;
+
+	rc = mdb_get(ldb->db_txn, ldb->db_dbi, &db_key, &db_data);
+	D_FREE(db_key.mv_data);
+	if (rc) {
+		rc = mdb_error2daos_error(rc);
+		goto out;
+	}
+
+	if (db_data.mv_size != val->iov_len) {
+		D_ERROR("mismatch value for table: %s, expected: %lu, got: %lu\n",
+			table, val->iov_len, db_data.mv_size);
+		rc = -DER_MISMATCH;
+		goto out;
+	}
+	memcpy(val->iov_buf, db_data.mv_data, db_data.mv_size);
+
+out:
+	if (end_tx) {
+		mdb_txn_abort(ldb->db_txn);
```

Comment:
Abort even fetch is successful?

Reply:
Yup, a bit strange. there are APIs:

mdb_txn_begin
mdb_txn_commit
mdb_txn_abort
mdb_txn_reset
mdb_txn_renew

No APIs like mdb_txn_end()..i search some example in the website, they used mdb_txn_abort() for read operations.